source/slang/slang-check-expr.cpp

```diff
@@ -795,8 +796,56 @@ namespace Slang
         return result;
     }
 
-    IntVal* SemanticsVisitor::TryConstantFoldExpr(
-        Expr* expr)
+    bool SemanticsVisitor::_checkForCircularityInConstantFolding(
+        Decl*                           decl,
+        ConstantFoldingCircularityInfo* circularityInfo)
```

Comment:
As an alternative it could use a bit in DeclCheckStateExt, or have some variable (as lots of unused bytes) in the decl that could be used to check without having to track a stack held linked list.

Reply:
That's true, although I'd like us to be slowly moving away from storing semantic-checking state on the AST nodes at all. Eventually the `DeclCheckState` for declarations should be stored in a `Dictionary<Decl*, DeclCheckStateExt>` owned by semantic checking, since it doesn't need to outlive the semantic checking step.

The state in this case is sort of local to the checking of a single declaration, so if we stored it globally (either as a global dictionary on the semantic checking context, or on the AST nodes themselves) we'd have to clear it out when we are done (in an RAII-like way that mirrors what we are constructing on the stack here anyway).